Refactor Device Plugin Registration

RenaudWasTaken opened the pull request:
**What this PR does / why we need it**:
This PR moves the storage responsibility away from the Manager struct and into an "Endpoint storage"
This allows for more extensive and reliable testing of the ReRegistration test introduced in PR #52290

This PR introduces multiple files and adds multiple tests:
- device_plugin_store.go (100% code coverage)
- endpoint.go (91.8% code coverage)
- endpoint_handler.go (95.8% code coverage)
- endpoint_store_shim.go (100% code coverage)

When paired with PR #52612 performance of the tests can get to <1s on a beefy machine.
Current test time usually takes at least 60s and more.
With this PR adding tests this reaches a whooping fixed 115s.

Note that most of the code not covered are error handling cases that can usually only be triggered
through changes that forgot to close or cleanup the structures.

This PR introduces an `EndpointHandler` interface that handles all interactions with the manager and abstracts away the complexity of handling registration and device usage:
```golang
type endpointHandler interface {
	NewEndpoint(resourceName, socketPath string) error

	Devices() map[string][]pluginapi.Device
	Endpoint(resourceName string) (endpoint, bool)

	Stop() error

	Store() endpointStore
	SetStore(endpointStore)

	SetCallback(f managerCallback)
}
```

![future arch](https://user-images.githubusercontent.com/2519571/34269352-4097a092-e684-11e7-88de-582a83946b43.png)


This allows for simple handling of complex behaviors such as the Re-registration bug:
```golang
func (h *endpointHandlerImpl) NewEndpoint(resourceName, socketPath string) error {
	e, err := newEndpoint(socketPath, resourceName)
	if err != nil {
		return err
	}

	// If an endpoint with this resourceName is already present
	var devStore deviceStore = newDeviceStoreImpl(h.callback)
	if old, ok := h.store.Endpoint(resourceName); ok && old != nil {
		devStore = old.Store()
	}

	// set a new store or the old store
	e.SetStore(devStore)

	// Add the endpoint to the store (and get the old one)
	old, ok := h.store.SwapEndpoint(e)
	go h.trackEnpoint(e)

	if ok && old != nil {
		// Prevent it to issue a Callback to the manager before stopping it
		old.SetStore(newAlwaysEmptyDeviceStore())

		if err = old.Stop(); err != nil {
			return err
		}
	}

	return nil
}
```

**Special notes for your reviewer**:
- @vishh @jiayingz @mindprince 
- Fixes #56026

**Release note**:
```release-note
NONE
```

xiangpengzhao commented:
/ok-to-test

RenaudWasTaken commented:
/test pull-kubernetes-unit

RenaudWasTaken commented:
/retest

RenaudWasTaken commented:
@jiayingz I believe I've fixed the tests can you take a look at this PR ?

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-etcd3

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-etcd3

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce

RenaudWasTaken commented:
/test pull-kubernetes-unit

jiayingz commented:
@RenaudWasTaken I took another look at this PR today and still feel we probably don't need to introduce another layer for device plugin endpoint today. Currently all the data managed by endpoint can be stored in a single map and its locking logic is not that complicated. It seems overkill to add an additional layer on top of endpoint to manage this data. I think we can revisit this idea if things have changed in the future that device plugin endpoint needs to manage more data states.

If the major motivation for this change to deflake the TestDevicePluginReRegistration unit test, I think the flakeness issue has already been fixed with PR https://github.com/kubernetes/kubernetes/pull/53028. It is just that PR wasn't cherry-picked into 1.8, so the test query returns some failed tests for ci-kubernetes-test-go-release-1.8 job. If we limit the query to the unit tests from head, we don't seem to have any recent failures from this test: https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=ci-kubernetes-test-go%24&test=TestDevicePluginReRegistration

I agree with you that it would be great if we can get rid of the sleeps in this unit test. However, as mentioned in the comment of TestDevicePluginReRegistration, the main thing this unit test wants to verify is that if a re-registration happens, we will NOT delete devices through unexpected callbacks. This test would lose its point if we remove this part of checking as happened in this PR.

I do like your idea of having a separate struct managing complicated data states. I think device plugin handler/manager will be a good place of applying this idea, as mentioned in issue https://github.com/kubernetes/kubernetes/issues/55180

RenaudWasTaken commented:
> If the major motivation for this change to deflake the TestDevicePluginReRegistration unit test, I think the flakeness issue has already been fixed with PR #53028. 

So the solution for de-flaking this test is to sleep 20 seconds. That doesn't seem like a good solution to me.

> It is just that PR wasn't cherry-picked into 1.8, so the test query returns some failed tests for ci-kubernetes-test-go-release-1.8 job. If we limit the query to the unit tests from head, we don't seem to have any recent failures from this test: https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=ci-kubernetes-test-go%24&test=TestDevicePluginReRegistration

If I run `go test -race -count 500 k8s.io/kubernetes/pkg/kubelet/deviceplugin -run
TestDevicePluginReRegistration -timeout 5h` on my machine it still fails quite often and it takes more than 5 hours.

That a unit test takes more than 5 hours (even if it's ran 500x) to run seems actually quite ridiculous.

> @RenaudWasTaken I took another look at this PR today and still feel we probably don't need to introduce another layer for device plugin endpoint today. Currently all the data managed by endpoint can be stored in a single map and its locking logic is not that complicated.

If you look at the code we're more introducing a concurrent map rather than an additional layer. Adding this `Store` or map will also allow us to swap it with a sync.Map as we get to golang 1.9.
Also if 

> It seems overkill to add an additional layer on top of endpoint to manage this data. I think we can revisit this idea if things have changed in the future that device plugin endpoint needs to manage more data states.

I'm sorry but what seems overkill to me is to have to wait 20 seconds in a unit test so that we can reduce the number of Flacks.
Essentially what we are saying by increasing the sleep duration rather than actually fixing the test is that we care more about not seeing the bug rather than the code being correct.

> I agree with you that it would be great if we can get rid of the sleeps in this unit test. However, as mentioned in the comment of TestDevicePluginReRegistration, the main thing this unit test wants to verify is that if a re-registration happens, we will NOT delete devices through unexpected callbacks. This test would lose its point if we remove this part of checking as happened in this PR.

If this is the acceptance criteria for this PR is to have a unit test that checks the number callbacks I'll add one.

RenaudWasTaken commented:
/area hw-accelerators

ConnorDoyle commented:
/assign @nqn

nqn commented on pkg/kubelet/cm/deviceplugin/endpoint_store.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+type store interface {
+	Get(resourceName string) (*endpoint, bool)
```
If this is a intermediate to be replaced by sync.Map, how about to let this return an `interface{}` instead?

RenaudWasTaken replied:
The idea is to replace the underlying implementation storage by a sync.Map (i.e: `endpointStore` here)  :)

That allows the type casting (and the error checks) to be done at the `endpointStore` level rather than handling that at the manager level.

vikaschoudhary16 replied:
Should this PR use sync.Map now straightway as @nqn shared on slack, https://groups.google.com/forum/#!topic/kubernetes-dev/INeuA9Flcb8

nqn commented on pkg/kubelet/cm/deviceplugin/endpoint_store.go:
```diff
@@ -0,0 +1,75 @@
+	}
+}
+
+func (s *endpointStore) Get(resourceName string) (*endpoint, bool) {
```
Think it's called `Load()` in https://github.com/golang/go/blob/master/src/sync/map.go#L101

nqn commented on pkg/kubelet/cm/deviceplugin/endpoint_store_stub.go:
```diff
@@ -0,0 +1,64 @@
+
+package deviceplugin
+
+type endpointStoreStub struct {
```
How is this a stub?
Do you mean shim? I may have a different understanding of stub in this context

RenaudWasTaken replied:
Agreed.
Was going for a stub at first I changed the model to a "channel-backed" shim in order to be able to
synchronously test the manager. 

Thanks for pointing that out!

nqn commented on pkg/kubelet/cm/deviceplugin/endpoint_store_stub.go:
```diff
@@ -0,0 +1,64 @@
+
+// ChanOut sends a message containing the parameters of the method called.
+// After sending that message, execution blocks until ChanContinue is written on
+func newInstrumentedEndpointStoreStub(outChan chan interface{}, continueChan chan bool) store {
```
I think I get it - so this is in the absence of a mocked interface to the endpoints?
First question would be if this is the idiomatic way of getting this done. Let's take a look around and see if this hasn't been solved before.

RenaudWasTaken replied:
I think you raise a good point about looking at how this is done :)

The idea behind this was that I wanted to check if at the moment an endpoint is persisted in storage, we advertise devices or not.
Meaning that if a device plugin re-registers we don't remove the number of devices it advertises during re-registration.

nqn reviewed (commented):
So, If I understand (and correct me if I am wrong):

1) We introduce the store to get rid of exposing locking at the call sites in endpoint.go. I like that and from reading the code would have suggested that as well. We can explore to get that to look more closely like the `sync.Map`, if that is what we are striving for replacing it with, when we can use Golang 1.9
2) With the endpoint store, we have a instrumented replacement. Called `endpointStoreStub` and `newInstrumentedEndpointStoreStub` which allows you to block or lock step in the test.

As a high level suggestion, if we define a more general state abstraction, like they did in the CPU manager, we may be able to have swappable in memory, file backed and mockable implementations. The latter can probably be defined inline in the test code. That way, you can close over channel variables in the test and do synchronization there.

I would miss some comments in the test about what is happening when, and why it works to block other threads from making progress.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/manager.go:
```diff
@@ -156,11 +152,9 @@ func (m *ManagerImpl) Allocate(resourceName string, devs []string) (*pluginapi.A
-	m.mutex.Lock()
-	e, ok := m.endpoints[resourceName]
-	m.mutex.Unlock()
+	glog.V(3).Infof("Recieved allocation request for devices %v for device plugin %s", devs, resourceName)
```
Thanks for noticing the typo!

Both are string + []string, so we won't go through go's mangling of the map. We can however replace `%v` by `%s` in the string case

vikaschoudhary16 reviewed (commented):
I think one more test case should be added where at reregistration devices, `devs`, is different from the `devs` at first registration.

Overall, i liked this channel based approach.

RenaudWasTaken commented:
@nqn @vikaschoudhary16 @jiayingz PTAL

- Increased test coverage by a lot in the manager part.
- Found a simple solution to device re-registration
- Test are not sleep dependent anymore
- Fixes a race https://github.com/kubernetes/kubernetes/issues/56026
- Endpoint `stop()` now guarantees that the endpoint is stopped when returning
- EndpointHandler `stop()` now guarantees that all endpoints are stopped and have been removed from the endpoint map

lichuqiang commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,202 @@
+		case <-syncChan:
+			continue
+		case <-time.After(time.Second):
+			return fmt.Errorf("Could not stop endpoint %s", s.e.resourceName)
```
is it better to continue and return an errorlist after the range finished?

RenaudWasTaken replied:
I don't think it really matters this error is really here to catch code changes (dependencies, ...) that might affect endpoint stopping.
If you encounter this error it means that the code needs to be changed.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,202 @@
+
+func (h *endpointHandlerImpl) NewEndpoint(resourceName, socketPath string) error {
+	var devStore deviceStore = newDeviceStoreImpl(h.callback)
+	if old, ok := h.store.Endpoint(resourceName); ok && old != nil {
```
Hmm I don't really agree here. `h.Endpoint` is a function exposed to anyone using the `endpointHandler` interface.

Here I am working with the struct's internals and know the underlying storage (The `EndpointStore` interface) and want to explicitly access it.

I don't expect `endpointHandler.Endpoint()` to mean anything else than `EndpointStore.Endpoint()` but here I am explicitly calling `EndpointStore.Endpoint()` (i.e: the endpoint storage).

RenaudWasTaken replied:
Feels mostly like coding style preference :)

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

tengqm reviewed (commented):
Overall looks great.
The PR title was confusing. I was incorrectly grouping this into test improvements. ;)

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_store.go:
```diff
@@ -0,0 +1,117 @@
+)
+
+type deviceStore interface {
+	Devices() []pluginapi.Device
```
I'm not sure what you mean by:
> The direct referencing of 'pluginapi.Device' looks like a shortcut to me.
> I'm not quite sure if we should combine Device and DeviceSpec into a representation internal to the store.

What are you trying to solve by combining `Device` and `DeviceSpec` ?

RenaudWasTaken replied:
> My rough feeling is that we will need to support more complicated use cases where the DeviceSpec must be taken into consideration.

Could you give an example of what you imagine might be a use case ?

tengqm replied:
GPU and CPU affinity for example.
We need to take some extended device information such as bus topology into consideration when allocating devices. A GPU device plugin itself is not supposed to be aware of the CPU status. IMHO, tasks like this are better handled at the host node or by the scheduler.
One way of doing this is to have kubelet "pre"-allocate all devices from device-plugin drivers. The device registry (store) has everything ready for a "real" allocation that comes later when a Pod requests this resource. Somehow, the Pod can express its need about CPU-GPU bandwidth, or maybe two GPUs on the same bus? Device manager then can do a better decision if it has the device specific information.
If the device registry (store) only caches the "health" status, we should reevaluate whether such a cache layer is worth the effort at all. Just some random thoughts.

RenaudWasTaken replied:
> We need to take some extended device information such as bus topology into consideration when allocating devices. A GPU device plugin itself is not supposed to be aware of the CPU status. IMHO, tasks like this are better handled at the host node or by the scheduler.

Agreed up to this point.

> One way of doing this is to have kubelet "pre"-allocate all devices from device-plugin drivers. The device registry (store) has everything ready for a "real" allocation that comes later when a Pod requests this resource. Somehow, the Pod can express its need about CPU-GPU bandwidth, or maybe two GPUs on the same bus? Device manager then can do a better decision if it has the device specific information.

I still don't understand how this loops back to having an internal representation that merges `Device` and `AllocateResponse`.

> If the device registry (store) only caches the "health" status, we should reevaluate whether such a cache layer is worth the effort at all. Just some random thoughts.
 
The device store is not a "cache layer" it's more of an abstraction for the concurrent device map that hides most of the locking logic as well as the "difference" logic.
It's a pretty useful abstraction when dealing with the re-registration bug.

lichuqiang replied:
>One way of doing this is to have kubelet "pre"-allocate all devices from device-plugin drivers. The device registry (store) has everything ready for a "real" allocation that comes later when a Pod requests this resource. Somehow, the Pod can express its need about CPU-GPU bandwidth, or maybe two GPUs on the same bus?

@tengqm  To support certain affinity in allocation is necessary, but handling the relationship entirely inside DevicePlugin might not be a good idea to some extent, maybe you could take a look at the doc of hardware topology:
https://docs.google.com/document/d/1dv3-se21ivEZhyRCr9qixvjyB_VRO2yIlotWb2kOdiw/edit?ts=5a135608#heading=h.t3uto9lxpu15
 and give your opinion there.

tengqm replied:
Thanks for the pointer. I was pondering over the overall picture. It doesn't belong to this PR.
However, device specs have to be stored somewhere to support this kind of advanced use cases.
An upper layer can combine different types of resources based on the details we stored here.
Currently these device specs are returned in AllocateResponse. We may need to revisit the design later. The focus at this stage is about the stable v1 iirc.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_store.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+type deviceStoreImpl struct {
+	sync.Mutex
```
We should have two goroutine accessing this at a time. One will be read another will be write.
Not sure an RWMutex makes a lot of sense as we don't have a lot of readers :)

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint.go:
```diff
@@ -95,91 +98,60 @@ func (e *endpointImpl) getDevices() []pluginapi.Device {
-		for _, d := range devs {
-			dOld, ok := devices[d.ID]
-			newDevs[d.ID] = d
+			s.Call(e.resourceName, nil, nil, s.Devices())
```
I'm not sure what you are referring to ?

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,224 @@
+
+func (h *endpointHandlerImpl) SetCallback(f monitorCallback) {
+	h.callback = f
+}
```
They are here because they are needed to access the internals of the objects during tests,

RenaudWasTaken replied:
To elaborate more they are needed in tests because they allow to stub the underlying internals.
So more of a good testing practice rather that over engineering.

Looking a bit more `Callback()` is never called so it can be removed.
`Range()` is a bit similar as it allows for simplification internally and for testing

tengqm replied:
I see.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,224 @@
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
+)
+
+type endpointHandler interface {
```
> Why are we introducing another layer of indirection?

More abstraction, more readability, more testability, more maintainability.
I don't think the current manager file which is composed of hundreds of lines is either one of the previous mentioned attributes.

The re-registration bug is a typical example of an architecture that does neither abstraction or testability. The current implementation solves this in a very simple way.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_store.go:
```diff
@@ -0,0 +1,117 @@
+	Update(devs []*pluginapi.Device) (added, updated, deleted []pluginapi.Device)
+
+	// Call calls the callback function
+	Call(resourceName string, added, updated, deleted []pluginapi.Device)
```
Sure sounds reasonable :)

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,224 @@
+	Store() endpointStore
+	SetStore(endpointStore)
+
+	Callback() monitorCallback
```
Sounds reasonable too :D

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/manager.go:
```diff
@@ -54,11 +54,10 @@ type ManagerImpl struct {
-
-	server *grpc.Server
+	// interface handling all Endpoint operations
+	ehdlr endpointHandler
```
Let's call this endpointHandler then

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,224 @@
+		devStore = old.Store()
+	}
+
+	e, err := newEndpoint(socketPath, resourceName, devStore)
```
Agreed, thanks for noticing this :)

RenaudWasTaken commented:
@tengqm thanks for the review.
The title underlies the aim of this PR which is to have a sane registration code.

I've updated the PR according to some of your comments.

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

RenaudWasTaken commented:
/test pull-kubernetes-unit

tengqm commented:
LGTM

tengqm commented:
@vishh @derekwaynecarr Please help take a look and approve if possible? This is for the most part a refactor work and it is paving a way for further code improvements.

RenaudWasTaken commented:
PR rebased.
Work is done, only waiting for more reviews :)

vikaschoudhary16 commented on pkg/kubelet/cm/deviceplugin/device_store.go:
```diff
@@ -0,0 +1,171 @@
+	HealthyDevices() []pluginapi.Device
+	Update(devs []*pluginapi.Device) (added, updated, deleted []pluginapi.Device)
+
+	// Call calls the callback function
```
You might want to add a better description comment.

vikaschoudhary16 reviewed (changes requested):
Please find few quick comments. I still need to take a deeper look.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/manager.go:
```diff
@@ -122,12 +118,15 @@ func newManagerImpl(socketPath string) (*ManagerImpl, error) {
 	}
 
+	manager.endpointHandler = newEndpointHandlerImpl(manager.genericDeviceUpdateCallback)
+
```
That's more of a coding style comment, and looks like we have no consistance on whether we prefer line returns before a return.

RenaudWasTaken commented:
@vikaschoudhary16 thanks for the review.

Looks like most of your comments are on me adding line returns to cleanly separate logical blocks in the code when adding code or rebasing so that I could get a better grasps on the code.

Since it looks like we have no clear consensus on whether we should have line returns after or before a mutex lock / unlock or a return let's not judge the value of the line return only on whether it adds to the diff or not.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/manager.go:
```diff
@@ -82,6 +77,8 @@ type ManagerImpl struct {
 	podDevices podDevices
 	store      utilstore.Store
+
+	server *grpc.Server
```
Mostly looks like a rebase artifact here

RenaudWasTaken replied:
This looked like a logical separation of the Manager's internal fields and the grpc.Server field

vikaschoudhary16 commented on pkg/kubelet/cm/deviceplugin/endpoint.go:
```diff
@@ -95,91 +102,60 @@ func (e *endpointImpl) getDevices() []pluginapi.Device {
+		return nil
+	}
+
+	select {
```
Should not it be `close(e.stopChan)` here?

RenaudWasTaken replied:
stopChan is closed by the `Run` function so that calling `Stop` means that all goroutine were terminated

vikaschoudhary16 commented on pkg/kubelet/cm/deviceplugin/endpoint.go:
```diff
@@ -95,91 +102,60 @@ func (e *endpointImpl) getDevices() []pluginapi.Device {
+	}
+
+	select {
+	case <-e.stopChan:
```
in what case, this `case` will be selected?

RenaudWasTaken replied:
See previous comment

vikaschoudhary16 reviewed (commented):
Other than few Nits here and there, changes LGTM mostly. 

This PR brings in a fair amount of complexity. IIUC, other than speeding up unit tests, this PR also solves #56026. #56026 can also be fixed by a relatively simpler change by copying endpoints.Devices over to plugin manager, instead of passing a reference of endpoint.Devices to plugin manager. 
Current PR adds complexity but brings in inflexibility and robustness by making it more pluggable and layered. 
I think lets all discuss if we should wait for the usecases, that would really benefit from this multi-layered abstraction, to become clear OR we already see clear benefits with these changes as of today.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,224 @@
+	SwapEndpoint(e endpoint) (endpoint, bool)
+	DeleteEndpoint(resourceName string) error
+
+	Range(func(string, *synchronizedEndpoint))
```
This was answered in another comment (hidden now):
- This is here here because this is needed to access the internals of the objects during tests.
- Range() is a bit similar as it allows for simplification internally and for testing

RenaudWasTaken commented:
> Current PR adds complexity but brings in inflexibility and robustness by making it more pluggable and layered.

Can you develop more on what you mean by "inflexibility" ? 

> IIUC, other than speeding up unit tests

Well paired with the next stub device plugin refactor we do get a test speedup.
I think one of the biggest value this PR brings is sane and better test coverage as well as sane behavior:
- No more time.Sleep are needed in tests
- Endpoints and Endpoint store now guarantee all goroutines were stopped once `Stop` returns
- More advanced tests can now be written

> I think lets all discuss if we should wait for the usecases, that would really benefit from this multi-layered abstraction, to become clear OR we already see clear benefits with these changes as of today.

If sane behavior, sane tests, better coverage, more robust code handling (see the re-registration bug) and performance improvement of x30 are not clear benefits please point me to the appropriate definition :)

vikaschoudhary16 commented:
> Can you develop more on what you mean by "inflexibility" ?

That was a typo. I meant `flexibility`. :)

vikaschoudhary16 commented:
@RenaudWasTaken #56026 is about the scenario where a device may disappear from endpoint store/database while manager has this device in its local cache. I dont think this PR resolves this issue.
Additionally, IMO, all these source code changes are just to improve unit test performance. I do not think it is a good idea to make source code overly complex just for the sake of test cases.

RenaudWasTaken commented:
> @RenaudWasTaken #56026 is about the scenario where a device may disappear from endpoint store/database while manager has this device in its local cache. I dont think this PR resolves this issue.

No that issue is about a real race where mutex locks are not handled correctly.

The scenario you are mentioning is a complex interaction that we always had and which we decided that this was not a problem because pods will be rejected if the device "disappears" / is reported as unhealthy either when the pod gets assigned to the node or when the Allocate call is made.

> Additionally, IMO, all these source code changes are just to improve unit test performance. I do not think it is a good idea to make source code overly complex just for the sake of test cases.

As mentioned multiple times, this PR does not improve test performance, only when paired with the stub device refactoring PR will it improve test performance.

What this PR does is bring sane behavior to tests and code, code modularity, and a testable architecture.

It's a commonly admitted that an architecture that you can't test is a pretty good indicator of a bad architecture.
A lot of the bugs we have are due to the fact that we can't correctly test the architecture. The re-registration bug is a typical example of this lack of testing ability:
- Multiple PRs to get the test right
- Multiple PRs to prevent test flacks on the CI (by increasing sleep time)
- Test time is insane (>60s on my 32 cores at 4Ghz builder) for absolutely no other reason than to have it pass on the CI when we can objectively get it down to < 1s
- Setup and Teardown of the tests are clearly too long indicating that either we are not unit testing or we have a separation problem
- Test correctness is objectively wrong because the current architecture offers no guarantee that the code has finished running when calling `Stop` and is only currently working because we `sleep` long enough

I believe this PR addresses if not entirely then at least in most part these issues.

> overly complex just for the sake of test cases.

Can you point out to actual complexity of the architecture.

My current impression is that the code brings a lot of abstraction that allows to focus only on the layer you are working with, making the code simpler to deal with.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/52616)
<!-- Reviewable:end -->

ScorpioCPH commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,228 @@
+	e.SetStore(devStore)
+
+	old, ok := h.store.SwapEndpoint(e)
+	go h.trackEnpoint(e)
```
Should we create new Endpoint if it is existed? Do you mean a new Endpoint with new `socketPath` but the same `resourceName`?

RenaudWasTaken replied:
If an endpoint already exists and we receive a new `Allocate` request for the same `resourceName` then we need to create a new endpoint to connect to that socket.

This is the current behavior and this PR does not change that.

ScorpioCPH commented on pkg/kubelet/cm/deviceplugin/endpoint.go:
```diff
@@ -33,11 +33,14 @@ import (
+	ResourceName() string
+
+	Store() deviceStore
+	SetStore(deviceStore)
```
Why we change these functions to be public?

RenaudWasTaken replied:
Answered above :)

ScorpioCPH reviewed (commented):
@RenaudWasTaken Thanks very much for your refactor works!

+1 for @vikaschoudhary16, IMHO, maybe this PR brings some complexity:

- As @tengqm mentioned: the PR title was a little confusing. This is a very big PR not just about test.
- Maybe we need some proposals or descriptions about this change, it is helpful for code review.
  - I saw more than one layer in your code, such as  deviceStore/endpointStore, it is a little difficult to figure out their relationships.
- There are some public functions/interfaces but not called out of this package `deviceplugin`, should them be private? 
- If we want improve test coverage, can we achieve this goal by adding more UT codes?

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_stub.go:
```diff
@@ -46,8 +48,9 @@ func NewDevicePluginStub(devs []*pluginapi.Device, socket string) *Stub {
-		stop:   make(chan interface{}),
-		update: make(chan []*pluginapi.Device),
+		stop:       make(chan interface{}),
+		registered: make(chan interface{}),
```
Well the whole stub device plugin is only for UT but yes

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_store.go:
```diff
@@ -0,0 +1,171 @@
+
+	// Testing methods
+	UpdateDevices(added, updated, deleted []pluginapi.Device)
+	UpdateAndCallback(resourceName string, added, updated, deleted []pluginapi.Device)
```
It's public in a private interface so it's not really exported.
Though if we do intend to export the types we can use them immediately.

So it has all the public pros with the private pros

RenaudWasTaken replied:
> Maybe we can move endpoint_* files into sub package, such as package endpoint?

Not sure it's a good idea I'm trying to keep the different layers in separate files.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/endpoint_handler.go:
```diff
@@ -0,0 +1,228 @@
+	go h.trackEnpoint(e)
+
+	if ok && old != nil {
+		old.SetStore(newAlwaysEmptyDeviceStore())
```
Endpoint now has a `DeviceStore` so that we can move the device map to the new Endpoint when re-registration happen.

The `DeviceStore` also has a Callback method that is Called when an update happens or when the endpoint is stoped.
In case of re-registration, we want to stop the old endpoint but not call the Callback, which is why we set the store to `newAlwaysEmptyDeviceStore`.
The endpoint will call the `DeviceStore` Callback but an `AlwaysEmptyDeviceStore` will never call the Manager callback.

RenaudWasTaken commented:
> +1 for @vikaschoudhary16, IMHO, maybe this PR brings some complexity:

@vikaschoudhary16 @ScorpioCPH Can I get some actual changes or reviews on what is complex?

At the moment my impression is that this is an unfamiliar code base that just "feels" complex for most people but they can't describe what is complex.

On the other hand the current code base is complex but known.
However I can point to actual complex, weird and incoherent behavior in the code (their are multiple links here).

I can also point to complex, insane and incorrect unit test as well as prove that with actual time metrics
as well as number of PRs that took to "fix" (and it's still not a correct test).

> As @tengqm mentioned: the PR title was a little confusing. This is a very big PR not just about test.
> Maybe we need some proposals or descriptions about this change, it is helpful for code review.
I saw more than one layer in your code, such as deviceStore/endpointStore, it is a little difficult to figure out their relationships.

I'll update the first post to address that lack of guidance.
The title points to what I think is an emblematic problem of the current architecture:
- An test that still at this day is not correct because the current architecture offers no guarantee that the code has finished running when calling Stop and is only currently working because we sleep long enough
- Multiple PRs to get the test right
- Multiple PRs to prevent test flacks on the CI (by increasing sleep time)
- Test time is insane (>60s on my 32 cores at 4Ghz builder) for absolutely no other reason than to have it pass on the CI when we can objectively get it down to < 1s
- Setup and Teardown of the tests are clearly too long indicating that either we are not unit testing or we have a separation problem

> There are some public functions/interfaces but not called out of this package deviceplugin, should them be private?

Addressed in comments.

> If we want improve test coverage, can we achieve this goal by adding more UT codes?

The goal of this PR is not to increase test coverage.

The UT test coverage increase is a side effect of having a better, saner, more flexible, more robust and more testable architecture introduced in this PR:
- More abstraction 
- More tests
- No more sleep
- Consistent behavior
- Strong guarantees on the code (`Stop` actually guarantees that all code stopped)

I'll add some more comments :)

RenaudWasTaken commented:
> +1 for @vikaschoudhary16, IMHO, maybe this PR brings some complexity:

@vikaschoudhary16 @ScorpioCPH Can I get some actual changes or reviews on what is complex?

At the moment my impression is that this is an unfamiliar code base that just "feels" complex for most people but they can't describe what is complex.

On the other hand the current code base is complex but known.
However I can point to actual [complex](https://github.com/kubernetes/kubernetes/blob/754bb1350faa34736de9216310dd00940bc56d8a/pkg/kubelet/cm/deviceplugin/manager.go#L314-L369) and [incoherent (stop gives no guarantee that the code is stopped)](https://github.com/kubernetes/kubernetes/blob/754bb1350faa34736de9216310dd00940bc56d8a/pkg/kubelet/cm/deviceplugin/endpoint.go#L181-L183) behavior in the code.

I can also point to [complex and incorrect](https://github.com/kubernetes/kubernetes/blob/754bb1350faa34736de9216310dd00940bc56d8a/pkg/kubelet/cm/deviceplugin/manager_test.go#L62-L126) unit test as well as prove that with actual time metrics as well as number of PRs that took to "fix" (and it's still not a correct test):
- 60 fixed seconds to run the whole test suite.
- When paired with PR #52612 < 1s on a strong machine <2s on the CI
- 4 PRs to get the re-registration test to stop causing flacks on the CI

> As @tengqm mentioned: the PR title was a little confusing. This is a very big PR not just about test.
> Maybe we need some proposals or descriptions about this change, it is helpful for code review.
I saw more than one layer in your code, such as deviceStore/endpointStore, it is a little difficult to figure out their relationships.

I'll update the first post to address that lack of guidance.
The title points to what I think is an emblematic problem of the current architecture:
- An test that still at this day is not correct because the current architecture offers no guarantee that the code has finished running when calling Stop and is only currently working because we sleep long enough
- Multiple PRs to get the test right
- Multiple PRs to prevent test flacks on the CI (by increasing sleep time)
- Test time is insane (>60s on my 32 cores at 4Ghz builder) for absolutely no other reason than to have it pass on the CI when we can objectively get it down to < 1s
- Setup and Teardown of the tests are clearly too long indicating that either we are not unit testing or we have a separation problem

At some point we have to realize that all these are red flags in terms of architecture quality. Especially when faced with an implementation that does not have these.

If they aren't I'd be interested to know what is.

> There are some public functions/interfaces but not called out of this package deviceplugin, should them be private?

Addressed in comments.

> If we want improve test coverage, can we achieve this goal by adding more UT codes?

The goal of this PR is not to increase test coverage.

The UT test coverage increase is a side effect of having a better, saner, more flexible, more robust and more testable architecture introduced in this PR:
- More abstraction 
- More tests
- No more sleep
- Consistent behavior
- Strong guarantees on the code (`Stop` actually guarantees that all code stopped)

I'll add some more comments :)

ConnorDoyle commented:
@RenaudWasTaken the PR description admits that there are multiple things going on here. It seems like this can be split into multiple smaller changes that are easier to review. For example, we could introduce the endpoint interface in one, change code to use it in a second, and rewrite tests in a third.

RenaudWasTaken commented:
@ConnorDoyle I could split that in multiple commits ?
Multiple PRs is going to be a pain :)

Commit:
Create stub device plugin directory on start

Commit:
Change stub device plugin stop order

Commit:
Add a device store for the device plugin

Commit:
Refactor endpoint to use the device store and guarantee all code stopped on Stop

Commit:
Added the device handler interface

RenaudWasTaken commented:
- Updated the first post
- Added more comments
- Multiple commits

tengqm commented:
/test pull-kubernetes-unit

tengqm commented:
@RenaudWasTaken thanks for the huge effort. I can see how this refactoring is helping improve the overall architecture deviceplugin manager. Suppose there are new unexpected behaviors from the device plugin (which is highly possible), or new proposals to polish the registration flow, we don't need to continue patching the DeviceManager layer. When we see device store flaky, we know where we need to look into.

There might still be nits, being it programming habits, being it interface names. We can always debate on those. At this point, I'd really hope we can focus on the blocking factors. If there are things unacceptable, we iterate on them, or else, we can get this in and move on.

RenaudWasTaken commented:
Thanks @tengqm ! I agree with you :)

Commit:
Update the manager and it's test to use the new device plugin handler interface

vikaschoudhary16 commented:
First of all, big changes are always prone to 'fixing some issues and opening some new issues'.

Now, here are my views on the merits/value which this PR introduces:
**1. Enhanced modularity and thus  flexibility by having number of interfaces:**
Interfaces/layers help when there could be more than one implementation. Interfaces help in smooth swap-in swap-out between possible alternative implementations. I am unable to think of multiple co-existing implementations of , for example, device store, which would benefit from these newly introduced abstractions. Would appreciate one or two concrete/specific examples that can help see the benefits of increased abstraction better. 

** 2. fixes #56026:**  
Issue description is too vague, unclear and too abstract to understand (at least for me). I would have written a unit test at least to establish that there is a race issue. I am not sure how this PR is fixing #56026 but would be happy to understand.  

** Re-Registration Scenario:** 
IMO, there was a race issue but we dont have that race anymore. PTAL at https://github.com/kubernetes/kubernetes/pull/57610 . Or there is still another issue which i am unaware of.

** Increased Unit Test Case Performance:**
This is not an apple-to-apple comparison. All test cases have been changed to using mocks. Today test cases use stub, which involves grpc calls just like actual plugin. Definitely using mock will be faster than using stub device plugin. 
More interestingly, `TestDevicePluginReRegistration` has totally been removed in this PR.

RenaudWasTaken commented:
> First of all, big changes are always prone to 'fixing some issues and opening some new issues'.

What's are your criterions for saying the current code base is not change friendly and what are your criterions for saying the new arch is better?

If I list my criterions they would be:
- Multiple PRs to get a change right
- Highly coupled code

Criterions for accepting would be:
- High test coverage
- Modular code

> Now, here are my views on the merits/value which this PR introduces:
> 1. Enhanced modularity and thus flexibility by having number of interfaces:
> Interfaces/layers help when there could be more than one implementation. Interfaces help in smooth swap-in swap-out between possible alternative implementations. I am unable to think of multiple co-existing implementations of , for example, device store, which would benefit from these newly introduced abstractions. Would appreciate one or two concrete/specific examples that can help see the benefits of increased abstraction better.

A concrete example is the current [registration logic](https://github.com/kubernetes/kubernetes/blob/f5f6f3e715cb8dfbd9657a4229c77ec6a5eab135/pkg/kubelet/cm/deviceplugin/manager.go#L314-L370) that took multiple times to get right.

Pretty sure that if we want to change the registration process later we'll have to deal with complex edge cases hard to change code and unreliable tests.

> 2. fixes #56026:
> Issue description is too vague, unclear and too abstract to understand (at least for me). I would have written a unit test at least to establish that there is a race issue. I am not sure how this PR is fixing #56026 but would be happy to understand.

I can add a test but as you probably know testing a race is not really a deterministic process.
You can take a look at the `device_store.go` file (I'll update the description).
I'm basically locking when computing the added, updated and removed lists.

> Re-Registration Scenario:
> IMO, there was a race issue but we dont have that race anymore. PTAL at #57610 . Or there is still another issue which i am unaware of.

The Re-registration test doesn't have a race. It's just plain incorrect.
A few example include the [20s wait](https://github.com/kubernetes/kubernetes/blob/f5f6f3e715cb8dfbd9657a4229c77ec6a5eab135/pkg/kubelet/cm/deviceplugin/manager_test.go#L89) for the callback to be issued.
Or the next [5s wait](https://github.com/kubernetes/kubernetes/blob/f5f6f3e715cb8dfbd9657a4229c77ec6a5eab135/pkg/kubelet/cm/deviceplugin/manager_test.go#L89) for the callback not to be issued.

> Increased Unit Test Case Performance:
> This is not an apple-to-apple comparison. All test cases have been changed to using mocks. Today test cases use stub, which involves grpc calls just like actual plugin. Definitely using mock will be faster than using stub device plugin.

Please take another look at the tests, that's not the case.
I've even doubled the time it takes to run the deviceplugin module from 60s to 120s.

> More interestingly, TestDevicePluginReRegistration has totally been removed in this PR.

Please take a look at the `TestReRegistration` test.

vikaschoudhary16 commented:
> If I list my criterions they would be:
> 
> Multiple PRs to get a change right

Thats how i think any software gets stabilized. Step by Step we find corner cases and fill the gaps.

> Highly coupled code

This is too generalized statement. I am still waiting for an example where coupling of components as it is in todays code, would not be able to handle the situation. 

> A concrete example is the current registration logic that took multiple times to get right.

Thats my point. We got it right now after a lot of efforts. Why to throw it away? Example you given is not the example. This is what you want to replace. Example i am interested is of a use case which makes you think that think current registration logic will not be able to handle.

> Pretty sure that if we want to change the registration process later we'll have to deal with complex edge cases

Thats what i am interested in example for. Do you have any such complex edge case example to share?


> I can add a test but as you probably know testing a race is not really a deterministic process.
> You can take a look at the device_store.go file (I'll update the description).
> I'm basically locking when computing the added, updated and removed lists.

If not test case, you may want to explain this race by running through steps like [i explained this previous race which is fixed now here 
](https://github.com/kubernetes/kubernetes/pull/57610#issue-284390981) .Something like,  these two threads would compete,  for this lock and how these threads would get triggered etc etc. 
You are saying that you are locking when computing the added,updated and removed list but not explaining clear enough that why this lock is necessary. Better would be to file a bug with proper dry-run kind of steps to explain what exactly will trigger a race in current code. You are stressing on solution without establishing problem.

> The Re-registration test doesn't have a race. It's just plain incorrect.
> A few example include the 20s wait for the callback to be issued.
> Or the next 5s wait for the callback not to be issued.

There was a race and because of that only sleeps were added. That race has got already fixed and these sleeps are not needed anymore. Please review https://github.com/kubernetes/kubernetes/pull/57610#issue-284390981 and share your concerns with removing sleep on that PR. 

Is there any other incorrectness in this test case?

> 
> It's basically saying that we don't care about test correctness we just care about not seing the bug. If the test was ran on a slower machine it would be failing.

Disagree, IMO, it will not fail now. Assuming that sleeps are not part of this test case anymore, Can you explain the flow which will make it fail?

RenaudWasTaken commented:
> Thats how i think any software gets stabilized. Step by Step we find corner cases and fill the gaps.

Glad we agree on that. This PR removes corner cases (usage of sleeps in tests) and fills the gaps of missing tests and abstraction.
I don't think increasing a sleep is a step towards stabilization, but rather a step towards ignoring a bug rather than fixing it.

> > Highly coupled code
> This is too generalized statement. I am still waiting for an example where coupling of components, as it is there in todays code, would not be able to handle the situation.

Sure if we re-write the registration logic as it is mentioned in a few discussion lately, I don't have any confidence in the test architecture to help us find bugs.
Additionally with this monolithic architecture we are violating a few good architecture principles such as the SRP.
It also reduces code dependency and makes test setup a lot more simple, thus increasing test quality.

In general I don't really want to debate highly coupled vs decoupled code. I think it's now a generally agreed design principle that the later is better and leads to better code.
And from personal experience as well as working in this code base I can confirm that the highly coupled architecture is a pain to deal with and that adding changes to the current architecture is not a walk in the park.

> Thats my point. We got it right now after a lot of efforts. Why to throw it away? Example you given is not the example. This is what you want to replace. Example that i am interested in is of a use case which makes you think that current registration logic will not be able to handle.

We don't have it right.
The current code does handle registration correctly but offers no guarantees with regards to code evolution. Just because "it works now" is not a good enough argument.

If we decide to have the registration code change, the current abstraction is clearly not good enough. Testing is pretty sparse and with the current code if the registration process takes long enough, the bug could still be present but pass the test successfully.

> If not test case, you may want to explain this race by running through steps like i explained this previous race which is fixed now here
> .Something like, these two threads would compete, for this lock and how these threads would get triggered etc etc.
> You are saying that you are locking when computing the added,updated and removed list but not explaining clear enough that why this lock is necessary. Better would be to file a bug with proper dry-run kind of steps to explain what exactly will trigger a race in current code. 

Race is pretty well explained in the issue.
Please read the issue if you want to understand what the race is.
Multiple people have contributed to make this race more understandable and propose different solutions. 

I'll point to the function fixing that race in the description.

> Quoting you:
> such as the Re-registration bug
> What is this Re-registration bug that you are referring here?

[Commented here](https://github.com/kubernetes/kubernetes/pull/57610#issuecomment-353869367)

> Disagree, IMO, it will not fail now. Assuming that sleeps are not part of this test case anymore, Can you explain the flow which will make it fail?

Yes it will still fail. [Assuming the current flow](https://github.com/kubernetes/kubernetes/blob/7dd82519da3f8c83d43efe18bebf500074a3c77c/pkg/kubelet/cm/deviceplugin/manager_test.go#L62) because the one you proposed is not correct.

Suppose the registration flow takes a bit longer or the machine this is running on is a bit slower and we change the flow to issue a callback the test will not catch this error.

> When paired with PR #52612 performance of the tests can get to <1s on a beefy machine
> This line seems to be suggesting that this PR is improving performance but then next line says that test case execution time got doubled in this PR. This is confusing.

Yes this PR adds more tests, leading to doubling the test time for this package.
When paired with PR #52612 the test time for this package gets under 1s.

Let's stop debating on what good programming practices we believe in or not. And try to get something actionable out of this conversation.
Please define what are criterions for accepting this PR or define what are the criterions that you think this PR should be rejected (and not "This PR brings in a fair amount of complexity").

I think that I've listed a fair amount of actionable criterions of why I think this PR brings a better architecture. We are currently going nowhere and I'd like to stop rebasing this indefinitely because we are fixing bugs of the old architecture that are already fixed here.

Either we accept it, change it or reject it but in either of these case we should have actionable items rather than a high level discussion on if we think highly coupled code is good or not.

vikaschoudhary16 commented:
I already said my points and disagreements are welcome. Lets have other's perspective as well.
